### PR TITLE
Add basic augmentations for segmentation models

### LINKED ISSUE #372: Enable better diagnostics: Example images
Example images are no longer written during segmentation model training. In the new PL training, this can only happen at rank zero, and means that we don't always get the same images coming in. Switch to doing it for the same subject across more ranks? Or always the first images on rank zero?
Check if `store_and_upload_example` is of any use, remove if not.
Tests should be updated: end of `_test_model_train()`

[AB#3909](https://innereye.visualstudio.com/60ce1777-00d6-4015-82bc-488a0c00202f/_workitems/edit/3909)

### PR DESCRIPTION
**Pull request by javier-alvarez**
Add config to enable disable basic augmentations for segmentation models
--apply_augmentations=True by default it is disabled to avoid regressions on previous models

closes #372 

With and without augmentation with 40 images. Not much improvement but a bit better on prostate and seminal vesicles.

jaalvare_augmentations/Run 49
Run 1: Last_experiment_baseline
2021-07-12T15:44:45Z INFO     Run 2: CURRENT
2021-07-12T15:44:45Z INFO     Name             N   Mean1   Mean2  N1>2 N2>1  Med1    Med2    WilcZ WilcP 2_vs_1
2021-07-12T15:44:45Z INFO     bladder         109  95.271  95.202   42   61  96.400  96.400  1.237 0.839 
2021-07-12T15:44:45Z INFO     external        109  99.339  99.332   11    4  99.500  99.500 -1.761 0.189 
2021-07-12T15:44:45Z INFO     femur_l         109  97.705  97.428   91    7  97.700  97.500 -7.907 0.000 WORSE
2021-07-12T15:44:45Z INFO     femur_r         109  97.554  97.570   43   45  97.700  97.700  0.191 0.561 
2021-07-12T15:44:45Z INFO     prostate        109  82.783  83.038   36   71  83.900  84.400  2.692 0.984 BETTER
2021-07-12T15:44:45Z INFO     rectum          109  84.951  84.724   57   51  86.400  86.300 -0.702 0.287 
2021-07-12T15:44:45Z INFO     seminalvesicles 109  70.379  71.193   45   62  72.600  73.100  1.908 0.937

### CODE REVIEW

File: InnerEye/ML/config.py [hunk 1]
```diff
@@ -788,9 +793,13 @@ def get_full_image_sample_transforms(self) -> ModelTransformsPerExecutionMode:
         """
         from InnerEye.ML.utils.transforms import Compose3D
         from InnerEye.ML.photometric_normalization import PhotometricNormalization
+        from InnerEye.ML.augmentations.augmentation_for_segmentation_utils import BasicAugmentations
 
         photometric_transformation = Compose3D(transforms=[PhotometricNormalization(self, use_gpu=False)])
-        return ModelTransformsPerExecutionMode(train=photometric_transformation,
+        training_augmentations = Compose3D(
+            transforms=[PhotometricNormalization(self, use_gpu=False), BasicAugmentations()])
+        training_transformation = training_augmentations if self.augmentations else photometric_transformation
```
Comment (ant0nsc) [inline InnerEye/ML/config.py:801]: DRY?
transforms= [Photometric()]
if self.augmentations:
    transforms.append(Basic..)
...Compose3D(transforms)

File: InnerEye/ML/config.py [hunk 2]
```diff
@@ -384,6 +384,9 @@ class SegmentationModelBase(ModelConfigBase):
     store_dataset_sample: bool = param.Boolean(False, doc="If True save image and segmentations for one image"
                                                           "in a batch for each training epoch")
 
+    #: If True apply basic 3D augmentations
+    augmentations: bool = param.Boolean(True, doc="If True apply basic 3D augmentations")
```
Comment (ant0nsc) [inline InnerEye/ML/config.py:388]: use a different name that hints at it being boolean? apply_augmentations?

File: InnerEye/ML/augmentations/augmentation_for_segmentation_utils.py [hunk 3]
```diff
@@ -126,3 +128,45 @@ def random_crop(sample: Sample,
...
+            image=tio.ScalarImage(tensor=image),
+            labels=tio.LabelMap(tensor=labels)
+        )
+        augment = tio.Compose([
+            tio.OneOf({
+                tio.RandomAffine(
+                    degrees=20),
+                tio.RandomNoise(),
+                tio.RandomMotion(),
+                tio.RandomBlur(),
```
Comment (ant0nsc) [inline InnerEye/ML/augmentations/augmentation_for_segmentation_utils.py:168]: Do noise, motion and blur have a visible effect on the images?

File: InnerEye/ML/augmentations/augmentation_for_segmentation_utils.py [hunk 4]
```diff
@@ -126,3 +128,45 @@ def random_crop(sample: Sample,
...
+            }, p=0.5)
+        ])
+        transformed_subject = augment(subject)
+        return transformed_subject.image.numpy(), transformed_subject.labels.numpy()
```
Comment (ant0nsc) [inline InnerEye/ML/augmentations/augmentation_for_segmentation_utils.py:172]: If you're applying "rotate", the labels also need to be rotated, is that the case now?

File: InnerEye/ML/config.py (hunk 1 above)
Comment (javier-alvarez) [reply InnerEye/ML/config.py:801]: for validation and test I dont want all the transforms so I need 2 lists

File: InnerEye/ML/augmentations/augmentation_for_segmentation_utils.py (hunk 3 above)
Comment (javier-alvarez) [reply InnerEye/ML/augmentations/augmentation_for_segmentation_utils.py:168]: Yes, I have a test. I will add tests next

File: InnerEye/ML/augmentations/augmentation_for_segmentation_utils.py (hunk 4 above)
Comment (javier-alvarez) [reply InnerEye/ML/augmentations/augmentation_for_segmentation_utils.py:172]: yes the torchio subject deals with this.

File: docs/building_models.md [hunk 5]
```diff
@@ -83,6 +83,8 @@ this case, you would simply omit the `azureml` flag, and instead of specifying
 `azure_dataset_id` in the class constructor, you can instead use `local_dataset="my/data/folder"`,
 where the folder `my/data/folder` contains a `dataset.csv` file and all the files that are referenced therein.
 
+If your dataset is small you can try `--apply_augmentations=True` for segmentation models this will apply `BasicAugmentations(Transform3D[Sample])`.
```
Comment (ant0nsc) [inline docs/building_models.md:86]: Would be good here to say which augmentations are actually done, and with which probabilities.

File: Tests/ML/augmentations/test_augmentation_for_segmentation_utils.py [hunk 6]
```diff
@@ -30,6 +33,60 @@
...
+        seg_path = f"{test_output_dirs.root_dir}/test_seg{i}.nii.gz"
+        io_util.store_as_nifti(transformed_sample.labels[0], image_with_header.header,
+                               file_name=seg_path, image_type=np.short)
+
+        """
+        Non-reproducible for some unknown reason. It would be good to uncomment once we know why        
+        expected_image_transform = load_nifti_image(full_ml_test_data_path(f"augmentation_baselines/test{i}.nii.gz"),
+                                                    np.short)
+        expected_seg_transform = load_nifti_image(full_ml_test_data_path(f"augmentation_baselines/test_seg{i}.nii.gz"),
+                                                  np.short)
```
Comment (ant0nsc) [inline Tests/ML/augmentations/test_augmentation_for_segmentation_utils.py:76]: These files should be checked in please, they are not part of the PR

**Review by ant0nsc — CHANGES_REQUESTED** [review]
